Handle bugs related to missing user DB entries

cmrd-senya opened the pull request:
fixes #7631

Commit:
Check for user existence on aspect membership destruction

cmrd-senya commented:
According to travis, the "improperly deleted users" can affect the account migration as well.

Commit:
Raise sensible error message when user is missing

That's for the case when podmin has messed up the database

Commit:
Forbid user destruction

cmrd-senya commented:
Looks like `User#destroy` was used as account deletion (close) method in the past. I wonder if we actually need a migration which removes the dangling contacts.

SuperTux88 commented on app/models/user.rb:
```diff
@@ -90,6 +90,10 @@ class User < ApplicationRecord
   after_save :remove_invalid_unconfirmed_emails
 
+  before_destroy do
+    raise "Never destroy users!"
```
+1 for this one, so podmins don't try to destroy the user via rails console.

SuperTux88 reviewed (commented):
> I wonder if we actually need a migration which removes the dangling contacts.

Yes, we should add a migration for that. And with a migration I'm not sure if we need the check anymore, because the user should always be there in the future ... but since the code is simpler now than before I think the check doesn't hurt ;)

SuperTux88 reviewed (changes requested):
@cmrd-senya can you add a migration to remove contacts with missing users? You can also remove contacts from other users with local people of missing users, then you can remove the commit "Drop contacts where respective user entry is missing".

Commit:
Clean up invalid contacts from the DB

Cleans invalid contacts where the referenced user was deleted from the DB or where
the referenced person owner was deleted from the DB.

cmrd-senya commented:
Updated

SuperTux88 reviewed (approved):
(wrong button ... and I can't delete reviews ...)

SuperTux88 reviewed (changes requested):
Can you please add migration models to the migration? Otherwise it looks good 👍 

(Turns out I have some missing users from 2011 too)

cmrd-senya commented:
Updated

SuperTux88 commented on db/migrate/20171012202650_cleanup_invalid_contacts.rb:
```diff
@@ -0,0 +1,41 @@
+    belongs_to :person
+
+    has_many :aspect_memberships, dependent: :destroy
+    has_many :aspects, through: :aspect_memberships
```
This line isn't needed

SuperTux88 commented on db/migrate/20171012202650_cleanup_invalid_contacts.rb:
```diff
@@ -0,0 +1,41 @@
+  end
+
+  class User < ApplicationRecord
+    has_one :person, inverse_of: :owner, foreign_key: :owner_id
```
Since you join from contacts via person to user, you should add the person model with the user relation. This line can be removed then, because it isn't needed.

SuperTux88 commented on db/migrate/20171012202650_cleanup_invalid_contacts.rb:
```diff
@@ -0,0 +1,41 @@
+  end
+
+  class Notification < ApplicationRecord
+    has_many :notification_actors, dependent: :destroy
```
Since `NotificationActor` doesn't do anything special on destroy, you can use `dependent: :delete_all` here, which is faster, because it doesn't instantiate the model (we could change that in the real model too).

cmrd-senya replied:
Maybe we can use `:delete_all` in `Contact` `:aspect_memberships` relation in migration model as well?

SuperTux88 replied:
Sure, that makes sense 👍

SuperTux88 reviewed (changes requested):
Only some small things :)

cmrd-senya commented:
Updated

SuperTux88 reviewed (changes requested):
Just tested it locally with my database and got an error:

```
Invalid single-table inheritance type: Notifications::StartedSharing is not a subclass of CleanupInvalidContacts::Notification
```

So it needs to also have this as a model:

```ruby
  module Notifications
    class StartedSharing < Notification
    end
  end
```

SuperTux88 commented:
OK, that doesn't work too :(

SuperTux88 commented:
OK, I did a different solution and pushed it directly to the PR:

```diff
diff --git a/db/migrate/20171012202650_cleanup_invalid_contacts.rb b/db/migrate/20171012202650_cleanup_invalid_contacts.rb
index c0b84c4b9..709724723 100644
--- a/db/migrate/20171012202650_cleanup_invalid_contacts.rb
+++ b/db/migrate/20171012202650_cleanup_invalid_contacts.rb
@@ -27,6 +27,7 @@ class CleanupInvalidContacts < ActiveRecord::Migration[5.1]
   end

   class Notification < ApplicationRecord
+    self.inheritance_column = nil
     has_many :notification_actors, dependent: :delete_all
   end
```
Polymorphic associations and single table inheritance always are bad with migration models :(

skyschub reviewed (approved):
Only approving @SuperTux88's changes, assuming the rest is reviewed by him.

SuperTux88 commented:
Yes, everything else looks good, so merged as 2e9c6f46dc6c01a63123445da876f23780b3398d, ad025850cece51672257929ab9e54b93f7fa70aa, 973e9d98c60226d216ecdd5874311e7b03e1445d and f212b51f7f83f63cb59dff45f38f82480bd7bb14

Thank you @cmrd-senya for the work 🍪 and thanks to @denschub for the quick review of the last fix 🍪
